コメント機能の強化（プロジェクトに対するコメント）

noriyotcp opened the pull request:
Closes #23 

既存のコメントを参考に以下のようにしている

![2018-08-02 20 51 20](https://user-images.githubusercontent.com/5820754/43581996-dc612f04-9695-11e8-9116-b209d615c84c.png)

- 既存の `comments` -> `card_comments` へのリネームは別issueにしたい

## 削除権限について
- 従来のコメントがユーザーが自分のコメントを削除できるようにはなっていない？カードの管理者は削除できるようだが

https://github.com/takeyuwebinc/gitfab2/blob/master/app/models/ability.rb#L72

それに倣い、`project_comments` も管理者のみが削除できるようにはしているが、要確認

## その他
- facebookコメント関連のコードは削除
- stl-viewに関する対応漏れがあったのでコードを削除

noriyotcp committed:
Remove unused code

Delete facebook comment

Remove stl-viewer from view and scss
- 以前stl_viewerを削除したが、その対応漏れ

https://github.com/takeyuwebinc/gitfab2/commit/63e04584d4aee69d454b997406f453767d0905df

noriyotcp committed:
Add ProjectComment

プロジェクト単位でのコメントを投稿・削除することができる

bodyはmaximum: 300

Add #is_project_manager? to User
プロジェクトを管理できる者がproject_commentを削除できる

Move project_comments/_index into recipes/_project_comments
Delete views/project_comments

Notify project owner(s) that a comment was posted

Project#manageable_by?(user)
- userがプロジェクトを管理できるかどうか

ProjectCommentsControllerでコメントの保存や削除に失敗した場合の処理を変更
- 以前はajaxに対応したものだったが、ajaxの廃止に伴い変更した

Move recipes/_project_comments to projects/_project_comments

Render 'projects/project_comments' in projects/show.html.slim

削除ボタンを表示する条件にcurrent_userの有無の評価を追加
- ユーザーがログアウトした状態でプロジェクトページにアクセスした際にエラーが出ていた

Change display format of project_comment's created_at
表示形式を2018-08-08の形式に変更
- プロジェクトの作成日時・更新日時の表示形式に合わせる

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+             project_comment: { body: body }
+           },
+           xhr: true
+    end
```
メソッドの生存期間はブロックに依存しないはずで、ブロック外からこのメソッドにアクセスできるが故にすごい悪さをし出すことがあるので、RSpecでブロック内にメソッドを作るのは絶対にやめてください。
こういう時は `subject` を使うのが普通です。もしくは、あまりやりたくないけど `shared_examples` 。

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+    let(:project) { FactoryBot.create(:project) }
+    let(:user) { FactoryBot.create(:user) }
+
+    def post_project_comment(project, body)
```
メソッド、作っちゃダメってわけじゃないですけれども、RSpec的には基本使いません。メモ化もされないし。
これは普通は `subject` でやります。

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+    context 'with valid parameters' do
+      before { sign_in user }
+
+      it 'should return a response ok(200)' do
```
RSpec2の名残かも知りませんが、 `should` いらない

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+      it 'should return a response ok(200)' do
+        post_project_comment(project, 'valid')
+        expect(response).to have_http_status(:ok)
```
`it 'hogehoge'` に書かれた文章とテストコードが見るからに同じ場合は `it { is_expected.to xx }` で書いた方がシンプルに見えますね。
そもそもRSpecのやりたいことが自然言語のように見えるコードを書くことなので、コードだけで十分自然言語らしくなっているのに文章を渡すとRSpecの存在意義が疑わしい。

```
subject { post :create, params: params }
context 'with valid params' do
  let(:params) { { foo: 'bar' } }
  it { is_expected.to have_http_status(:ok) }
end
```

itkrt2y replied:
あと、読む側に無駄な思考コストをかけさせないため、とにかくブロックの中からブロック見出しに書かれたものと関わりのないロジックを排除したいです。

例えば、
- `describe 'POST #create'` の次の行に `subject { post :create } ` を書く
- `context 'with valid params'` の次の行に `let(:params) { { body: 'valid' } } `を書く
- `context 'with invalid params'` の次の行に `let(:params) { { body: nil } } `を書く
- `it "creates a ProjectComment"` の次の行に `it { expect{ subject }.to change{ ProjectComment.count }.by(1)` を書く（これはdescriptionなくてもいいけど）

という具合に、見出しの後にそのブロックの本題を書き、また、どっかで定義されたメソッドの内部実装なんかを意識しないでもその部分だけ読めば何がしたくてどうしてそのexpectedになるかを理解できるように書きたいです。テストはコードジャンプの数だけとにかく読みづらくなる。

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    if @project_comment.save
+      render :index
+    else
+      @message = @project_comment.errors.messages || { "Error:": "Something's wrong" }
```
`@project_comment.errors.messages` => `@project_comment.errors.full_messages`

itkrt2y replied:
あと、インスタンス変数の必要はない

itkrt2y commented on app/views/project_comments/index.js.erb:
```diff
@@ -0,0 +1,2 @@
+document.getElementById("project_comments_area").innerHTML = "<%= j(render 'index', { project_comments: @project_comment.project.project_comments }) %>"
```
「直せ」ってわけではないですが、この `.js.erb` を返すやり方が今時よろしくないのはわかると思うのですが、何か改善方法を考えましたか？

itkrt2y commented on app/views/project_comments/_form.html.slim:
```diff
@@ -0,0 +1,7 @@
+    .left
+      = image_tag current_user.avatar.thumb.url, class: "avatar"
+    .right
+      = f.text_area :body, maxlength: 300, size: "40x4", class: "comment-editor"
```
formにmaxlengthを指定するのは、例えば指定文字数以上の文章をコピペした時とかに勝手に切り落とされるけれどもユーザーは気づかず送信してしまうみたいなことが起こるので、どちらかといえばアンチパターンです。
文字数制限をするなら、現在の入力文字数のカウント表示とバリデーションエラーが基本。文字数カウント表示までは依頼にないしJSを今は必要以上汚したくないので、バリデーションエラーがあれば十分。

itkrt2y replied:
そもそも300文字という制限がどこから来たのかよくわかりませんが...

noriyotcp replied:
元々は`components/_card_comments.html.slim`から由来しています

https://github.com/takeyuwebinc/gitfab2/blob/master/app/views/components/_card_comments.html.slim#L17

今確認したらそちらのほうも文字数表示が無いですね・・・

itkrt2y commented on app/views/project_comments/_index.html.slim:
```diff
@@ -0,0 +1,23 @@
+ul.comments
+  - project_comments.each_with_index do |comment, i|
```
`each.with_index(1)`

itkrt2y commented on app/views/recipes/_project_comments.html.slim:
```diff
@@ -0,0 +1,7 @@
+    .comments-wrapper
+      #project_comments_area
+        = render partial: 'project_comments/index', locals: { project_comments: @project_comments }
+      = render partial: 'project_comments/form', locals: { project_comment: @project_comment, project: @project }
```
なんでこんなにpartial分けてるんです？

itkrt2y replied:
あと、partialの中の `- if current_user` はこの `render partial:` に被せるように存在すべき

itkrt2y commented on app/models/ability.rb:
```diff
@@ -71,6 +71,12 @@ def initialize(user, params)
+    end
+    can :destroy, ProjectComment do |comment|
+      can? :manage, comment.project
+    end
```
個人的都合ですが、cancancanは捨てたいのでこれからはこの手のロジックはModelに書いて欲しいです

itkrt2y commented on app/controllers/projects_controller.rb:
```diff
@@ -19,6 +19,8 @@ def show
     not_found if @project.blank?
 
     @recipe = @project.recipe
+    @project_comment = ProjectComment.new
```
フォーム用のインスタンス変数は必要ですか？
`form_with` の送信内容に `project_id` と `current_user.id` の2つを含められれば、 `ProjectComment` のインスタンス変数は不要に思えます。

itkrt2y commented on app/views/project_comments/_form.html.slim:
```diff
@@ -0,0 +1,7 @@
+- if current_user
+  = form_with(scope: project_comment, url: project_project_comments_path(project.owner, project.id), class: 'comment-form') do |f|
```
`form_with` は `.right` の下で良さそう

itkrt2y commented:
コメント表示、正直ajaxにする必然性はないと思っています。
コメント投稿したら、URLにコメントへのアンカーをつけてリダイレクトで十分かと。

最高のUXを求めるなら確かにリロードなしに反映したいけど、フロントエンドが無秩序な現状にさらにコードを追加するのはなるべく避けたい。

noriyotcp commented:
540bf3e
Ajaxをやめてアンカーつけてリダイレクトするようにしました

投稿したあと
![2018-08-08 19 10 45](https://user-images.githubusercontent.com/5820754/43832019-fc422990-9b40-11e8-858b-bd3990cd8467.png)

コメント削除したあとは`section#project-comments` へリダイレクト
![2018-08-08 19 11 13](https://user-images.githubusercontent.com/5820754/43832050-12f2f3ea-9b41-11e8-913d-22b60eed4a76.png)

ProjectComment一覧も`recipes/_project_comments.html.slim` に集約することにしました

noriyotcp commented:
あ、既存のコメントに合わせるならば、投稿後の通知機能をまだ実装してなかった

noriyotcp commented:
Ajaxの廃止と通知機能の実装をしました

itkrt2y commented on app/models/project.rb:
```diff
@@ -193,6 +194,10 @@ def update_draft!
     update!(draft: generate_draft)
   end
 
+  def managed_by?(user)
```
引数のuserがprojectを操作可能か聞きたいのだと思うのですが、"managed by user?"では「userに操作された？』と過去に引数のユーザーによって操作されたことがあるか聞いてるように見えます。
`manageable_by?(user)` の方がそれらしい。

noriyotcp replied:
(Is the project) managed by user? のつもりでしたが、かといって `is_managed_by?(user)`  も少し冗長だし、`manageable_by?(user)` にします 👍

itkrt2y replied:
"project is managed by user" でも「プロジェクトはユーザーに操作された」で最初のコメントと変わらないと思いますが......

noriyotcp replied:
あ、操作ではなくて「管理」そして現在形で「プロジェクトはユーザーに管理されている」というつもりでした
そしてそれが真なら、ユーザーがプロジェクトのコメントを操作できる

なんにせよ`manageable_by?` のほうがハッキリしていて良いですね

itkrt2y replied:
なるほど

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,125 @@
+      it do
+        expect {
+          subject
+        }.to change(ProjectComment, :count).by(1)
```
好みの範囲ですけど、一行にして欲しい。文章として "it expect to change"と読みたい。

これ
    だと
なんか変な改行の
         文章を読んでる気分

itkrt2y replied:
好みの範囲ではありますが、ワンライナーで書いて欲しいです。
これ
&nbsp;&nbsp;だと
&nbsp;&nbsp;&nbsp;&nbsp;なんか変な
&nbsp;&nbsp;改行の文章を読んでる気分


```rb
it { expect{ subject }.to change(ProjectComment, :count).by(1) }
```

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,125 @@
+    it { expect(subject.notified_id).to eq project.owner.id }
+    it { expect(subject.notificatable_url).to eq project_path(project.owner.name, project) }
+    it { expect(subject.notificatable_type).to eq 'Project' }
+    it { expect(subject.body).to eq "#{user.name} commented on #{project.title}." }
```
```rb
it do
  is_expected.to have_attributes(
    notifier_id: user.id,
    notified_id: project.owner.id,
    notificatable_url: project_path(project.owner.name, project),
    # 中略
  )
end
```

itkrt2y commented on spec/models/user_spec.rb:
```diff
@@ -122,6 +122,42 @@
+
+    context 'When the user is admin of the project' do
+      let(:project) { FactoryBot.create(:group_project) }
+      before { allow_any_instance_of(User).to receive(:is_admin_of?).and_return(true) }
```
`allow(user).to receive(:is_admin_of?).and_return(true)`

対象を制限できるのならば、それに越したことはない。

itkrt2y commented on spec/models/project_spec.rb:
```diff
@@ -221,4 +221,23 @@
+    subject { project.managed_by?(FactoryBot.create(:user)) }
+
+    context 'When the project is deleted' do
+      before { project.update!(is_deleted: true) }
```
`describe '#managed_by?'` が外の状態に依存しているの、RSpec的と言えば確かにそうですが、 `before update` とかし始めると外部の知識に依存しすぎている ように思えます。
ぱっと見どこで定義されたかわからないものを更新するくらいならば、新しく `let(:project)` を作る方がいい。

個人的意見としては、メソッドごとに切り離してクラスで共有される状態は無くしたい。

```rb
describe '#managed_by?' do
  subject { project.managed_by?(FactoryBot.create(:user)) }

  let(:project) { FactoryBot.create(:project, owner: user, is_deleted: is_deleted }
  let(:user) { FactoryBot.create(:user) }
  let(:is_deleted) { false }

  context 'when the project is deleted' do
    let(:is_deleted) { true }
    it { is_expected.to be false }
  end

  context 'when user can manage a project' do
    before { allow(user).to receive(:is_project_manager?).and_return(true) }
    it { is_expected.to be true }
  end
end
```

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,40 @@
+    @project_comment = ProjectComment.find(params[:id])
+
+    if @project_comment.destroy
+      redirect_to project_path(project.owner.name, project, anchor: "project-comments")
```
`project#id` が欲しいだけなら `Project.find` をする必要はなくて `project_path` に `params[:project_id]` を渡せばいい。

noriyotcp replied:
idというかslugで投げないといけないですね

`http://test.host/user-hoge/project-hoge#project-comments`

itkrt2y replied:
じゃあslugだけ取得したいですね。ARオブジェクト作るのはどちらにしろ無駄

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,40 @@
+    project = Project.find(params[:project_id])
+    @project_comment = ProjectComment.find(params[:id])
+
+    if @project_comment.destroy
```
`current_user` がコメントを削除可能か確認する必要があります。

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,40 @@
+      redirect_to project_path(project.owner.name, project, anchor: "project-comment-#{@project_comment.id}")
+    else
+      message = @project_comment.errors.messages || { "Error:": "Something's wrong" }
+      render json: { success: false, message: message }, status: 400
```
コメント作成失敗側がajax処理の時のままですね

itkrt2y replied:
`destory` のも

itkrt2y commented on app/views/recipes/_project_comments.html.slim:
```diff
@@ -0,0 +1,34 @@
+                  span.created-at.header-content
+                    = "Posted date: "
+                    span.date
+                      = comment.created_at.strftime '%Y.%m.%d.%H.%M'
```
2点

1. viewでは `strftime` するよりは、 `l(comment.created_at)` と書いた方がシンプル（ https://guides.rubyonrails.org/i18n.html#adding-date-time-formats ）
1. わかってて依頼がないのでスルーしてるところですが、TimeZone設定がされてないのでこれだとUTCで表示される

多分年月日表記までで、時刻なくていいです

itkrt2y commented on app/controllers/projects_controller.rb:
```diff
@@ -19,6 +19,7 @@ def show
     not_found if @project.blank?
 
     @recipe = @project.recipe
+    @project_comments = @project.project_comments
```
`order(:id)` ないとダメですね。SQLの取得値は順番が保証されない。

itkrt2y commented on app/views/recipes/_project_comments.html.slim:
```diff
@@ -0,0 +1,34 @@
+section#project-comments
```
`app/views/recipes/` ディレクトリを消したので、適宜読み替えてください。
多分 `app/views/projects/` になると思います

itkrt2y commented:
`app/views/recipes/` ディレクトリを消したので、適宜書き換えてください。
多分 `app/views/projects/` になると思います

itkrt2y commented:
`comments` テーブルを `card_comments` にリネームするのは、他のタスクの関係上、僕の方でやりました https://github.com/takeyuwebinc/gitfab2/commit/bb1310aa35a1140acba9499bafca716eaeff8fe9

itkrt2y commented:
commit、squashできませんか？

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    @project_comment = ProjectComment.find(params[:id])
+    project = @project_comment.project
+
+    return unless project.manageable_by?(current_user)
```
`redirect` か `render` する

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    else
+      redirect_to project_path(project.owner.name, project, anchor: "project-comment-form"),
+                  alert: @project_comment.errors.full_messages,
+                  flash: { project_comment_body: @project_comment.body }
```
`anchor` とかで気持ちは理解しますけど、エラー発生時に入力されたコメントが消えちゃいますし `redirect` はダメでは

itkrt2y replied:
あと、もし必ず `redirect` するのならば、このactionではインスタンス変数は不要

noriyotcp replied:
コメントはflash[:project_comment_body] に格納して、ここで表示させるようにしていました

https://github.com/takeyuwebinc/gitfab2/pull/37/commits/b4ca4f5273b7968a609d390697eabbab58809ce6#diff-f158e2a4e9169f410f5f0f7b4ca2bec1R38

![2018-08-10 17 03 28](https://user-images.githubusercontent.com/5820754/44026022-aa6400e0-9f2c-11e8-82f8-8f72b45636fe.png)

itkrt2y replied:
あー、まあ、なくはないか......

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,139 @@
+
+    it { expect{ subject }.to change(ProjectComment, :count).by(-1) }
+
+    context 'when the comment could not be deleted' do
```
なんか妙な表現ですね。 `when user could not delete a comment` くらいで良さそう

itkrt2y commented on config/application.rb:
```diff
@@ -11,6 +11,8 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 5.2
 
+    config.time_zone = 'Tokyo'
```
2点

1. やるなら別PRで
2. 既存データの日時と整合性が取れることは確認済みですか？

noriyotcp replied:
整合性が取れる、というのは確かに既存の`created_at` の日時に＋９時間されている、ということでしょうか？

```sh
ProjectComment.last
=> #<ProjectComment (snip) created_at: "2018-08-10 08:52:48", updated_at: "2018-08-10 08:52:48">
> ProjectComment.last.created_at
  ProjectComment Load (1.1ms)  SELECT  `project_comments`.* FROM `project_comments` ORDER BY `project_comments`.`id` DESC LIMIT 1
=> Fri, 10 Aug 2018 17:52:48 JST +09:00
```

itkrt2y replied:
- 既存処理でJSTな時間をUTC扱いで保存しているものがないか（タイムゾーンはUTCなのに+9時間された値が保存されてないか）
    - 時刻を入力するフォームが存在した場合などが危ない。例えばユーザーが「13時」と入力すればそれは「JSTの13時」を示しているだろうが、タイムゾーン設定をしていないのでデータベースはそれを「UTCの13時」として保存することになる
- DateTimeはUTCで保存されるべきだろうが、Dateは逆にJSTな日付で保存されてるべきだろう。9時間ずれることで既存データで日付がずれて見えるものが出てくると思うが、それは大丈夫か
    - 例： UTC13日21時/JST14日6時でデータを保存する際、これまでDate型には13日が保存されていたが、タイムゾーンを変更すると14日が保存されるようになる。これまでのデータにこの1日ズレが存在してもロジック上に問題ないのか。
    - と、ここまで書いたところで `schema.rb` を見たらDate型はなかったから大丈夫だと思われるが、一応、観点の共有として

itkrt2y commented on spec/models/project_spec.rb:
```diff
@@ -233,4 +233,27 @@
+      it { is_expected.to be false }
+    end
+
+    context 'When the project is managed by the user' do
```
メソッド名を変更したのを反映していない。
メソッド名に合わせるか、合わせないでもわかりやすくするかした方が良いですね。

僕だったら `when user is a project manager` にしちゃいます

noriyotcp committed:
Redirect when current_user is not a project manager

Change instance variables to local variables
- ProjectComments#create, destroyどちらのアクションでも必ずリダイレクトするのでインスタンス変数が不要

Refactor docstring in spec

noriyotcp commented:
コミットをまとめました。最後のコミットが直近の指摘事項に対する修正となります

タイムゾーンの設定は削除しました

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+    if project_comment.save
+      notify_users(project)
+      redirect_to project_path(project.owner.name, project, anchor: "project-comment-#{project_comment.id}")
```
routesに変更入れて、 `project_path` に `owner_name` を渡さないようになりました。
https://github.com/takeyuwebinc/gitfab2/commit/fba61f60ef78945d28dc77c0c16f8a34cb1fe3b6

今回だと以下のように1個目の引数の `project.owner.name` がいらないです

```rb
redirect_to project_path(project, anchor: "project-comment-#{project_comment.id}")
```

これによって意識しないでよくなったことですが、レビューしてて気づいたこと。
`owner_name` に渡す値は `owner.name` じゃなくて `owner.slug` の方がいいです。

itkrt2y replied:
大変申し訳ないですが、これrevertします

noriyotcp committed:
Pass project.owner.slug instead of project.owner.name

noriyotcp commented on app/assets/stylesheets/projects/project_comments.scss:
```diff
@@ -0,0 +1,29 @@
+  margin-bottom: 50px;
+  margin-left: $first_column_width;
+  padding-top: 1px;
+  width: $article_column_width;
```
> Name of variable `article_column_width` should be written in all lowercase letters with hyphens instead of underscores
> [link](https://sider.review/gh/71619232/pull_requests/37)

これは影響範囲が大きい。やるなら別PRにしたい

noriyotcp commented on app/assets/stylesheets/projects/project_comments.scss:
```diff
@@ -0,0 +1,29 @@
+@import 'mixins/clearfix';
+@import 'projects/project_variables';
+
+section#project-comments {
```
> Avoid using id selectors
> [link](https://sider.review/gh/71619232/pull_requests/37)

気持ちはわかるが、HTML側も変更を要するので現段階ではクローズしたい

itkrt2y replied:
scss-lintの設定ファイル書いていいですよ

noriyotcp commented:
2fb32b2 slugを投げる形に変更しました

itkrt2y commented on .rubocop.yml:
```diff
@@ -73,6 +73,9 @@ PredicateName:
     - hoge_
 
+Style/AndOr:
+  Enabled: false
```
これ、 `redirect_to hoge and return` のところだと思いますけど、これを無効にするのではなくて、普通に `return` を改行して書いて欲しいです。

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+    unless project.manageable_by?(current_user)
+      redirect_to project_path(project.owner.slug, project, anchor: "project-comments"),
+                  alert: 'You can not delete a comment' and return
```
rubocopのStyle/AndOrはここですよね。
こんなところにreturn置かれてもさすがに読みづらいだけ。改行するべき

itkrt2y commented on app/controllers/project_comments_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+    if project_comment.save
+      notify_users(project)
+      redirect_to project_path(project.owner.slug, project, anchor: "project-comment-#{project_comment.id}")
```
直さなくてもいいですが、friendly_idのおかげなのかARオブジェクトを渡せば勝手にslugを使ってくれます

```rb
# これでslugを使ったpathが生成される
project_path(project.owner, project)
```

itkrt2y commented on Gemfile:
```diff
@@ -40,6 +40,7 @@ group :development, :test do
   gem 'factory_bot_rails'
   gem 'rspec-rails'
   gem 'rubocop', require: false
+  gem 'scss_lint', require: false
```
slim_lintもですけど、Gemfileに含める必要ありますかね？
使いたい人が `gem install scss_lint` して使えばいいのではと思うのですが。

noriyotcp replied:
あー手元で実行して確認できるといいかな、と思ったのですが、各自`gem install` で十分ですね

itkrt2y commented on spec/controllers/project_comments_controller_spec.rb:
```diff
@@ -0,0 +1,146 @@
+
+    before do
+      sign_in project.owner
+      project_comment
```
必ずbeforeでこれを呼び出すなら `let!` 使えばいいのでは

noriyotcp committed:
Add .scss-lint.yml and remove slim_lint gem

project_comments_controller.rb
and returnではなく改行してreturn

projects/project_comments.scss
partialをインポートする際はunderscoreをトル

noriyotcp committed:
Improve project_comments_controller_spec

project_commentの生成をbefore内からlet!へ変更

factories/project_comments

FactoryBot/AttributeDefinedStaticallyの修正
https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/FactoryBot/AttributeDefinedStatically

noriyotcp commented:
scss_lint, slim_lint gem の削除と、specの修正を行いました

itkrt2y commented:
マージしてから気づきましたが、これ、自分のコメントを削除できませんね

itkrt2y commented:
僕の方で直し中

noriyotcp commented:
すみません！
旧Comment(現・CardComment) も自分のコメント削除できないのでそれに合わせたんですが、そもそも実装が間違ってて現在は修正済みなんですね 🙏 

https://github.com/takeyuwebinc/gitfab2/commit/6155d3297000ecba19284f73f0032eda4686e127#diff-e9c04d21b8c4d777e4231c4c6ce5f8cb

itkrt2y commented:
文字数多いと無限に領域外に伸びていく...... （動作テストサボったのが行けないですね）

noriyotcp commented:
うああああほんとだ

![2018-08-20 18 51 58](https://user-images.githubusercontent.com/5820754/44333613-3b3c4000-a4aa-11e8-933b-86de062fb7ab.png)

noriyotcp commented:
あれ、これ既存の`CardComment` のほうでも同じなんですね

![2018-08-20 18 55 05](https://user-images.githubusercontent.com/5820754/44333728-a554e500-a4aa-11e8-86b0-b188df51ad8d.png)

itkrt2y commented:
直した

itkrt2y commented:
既存のものは基本的にバグりまくりなので参考にならないですし、気付き次第潰すしかないです
